_posts/2019-04-10-imperatywnie-vs-funkcyjnie.md

```diff
@@ -0,0 +1,118 @@
+
+```kotlin
+fun multiply(a: Int, b: Int) = a * b.toFloat()
+fun divide(a: Int, b: Int) = a / b.toFloat()
+ 
+multiply(123456789, 123456789) // 1.52415794E16
+divide(123456789, 0) // Infinity
+
+result.isInfinite().maybe { print("You should learn the basics of math, probably.") }
+```
+
+W tym momencie dla każdej możliwej wartości będzie ten sam wynik. Również w przypadku poprzedniego wyjątku teraz 
+zostanie zwrócony obiekt **Infinity**, na którym możemy działać dalej z naszym kodem. Można użyć extension function z Kotlina, które pozwala rozszerzyć każdą metodę o dodatkowe funkcjonalności. W tym przypadku zaciągnęliśmy biblioteką Arrow, która zrobiła za nas trochę magii `maybe { ... }`, które zwraca `none()` dla braku wartości lub `some()` jeśli coś jest. W Javie niby też mamy obiekty, ale Kotlin poszedł krok dalej, bo dosłownie wszystko obiektem. Tak też zamiast `void` jest `Unit`, a gdy wiemy, że metoda zawsze się nie udaje to jest `Nothing`. Jest to przydatne chociażby w metodzie `TODO("")` która zwraca właśnie `Nothing`. Jest jeszcze `Any` dla wartości non-nullable oraz `Any?` dla nullable co ma swój odpowiednik w Javowym `Object`. Na tym zakończmy te rozważania, bo jest to poza zakresem wpisu. Potraktujmy to jako ciekawostkę, bo rzeczy te przydają się również w programowaniu funkcyjnym.
+
+Gdy wiemy czego się spodziewać po funkcji i zawsze zwraca ona prawidłową wartość można nazwać ją czystą. Znaczy to tyle, że takie pure functions nie mają żadnych efektów ubocznych. To co zwraca zależy tylko i wyłącznie od parametrów jakie podaliśmy. Dobrym przykładem są funkcje z klasy Math. Gdzie biorąc `sqrt(2.0)` wiemy, że wynikiem jest zawsze jakiś obiekt typu Double (jak było wcześniej powiedziane w Kotlinie wszystko jest obiektem również prymitywy, których bądź co bądź po prostu nie ma).
+
+## To jak pisać bardziej funkcyjnie?
+
+```kotlin
+// Na początek imperatywny przykład
+fun buyBook(bookName: String, creditCard: CreditCard): Book {
+    val book = Book(name = bookName)
+    creditCard.performPayment(book.price)
+    return book
+}
+```
+
+Powyższy kod jest trudny do przetestowania. Musielibyśmy sobie zamockować proces weryfikacji w banku, czy karta w ogóle istnieje, czy posiada środki na koncie i tak dalej. Dopiero po tym zwracamy film. Zróbmy to trochę inaczej. Co jeśliby przetestować to bez kontaktu z bankiem? Moglibyśmy powiązać płatność z książką.
```

Comment:
niezbyt mi się to podoba... nie skupiaj się na testowaniu. Może lepiej byłoby napisać, że nie możemy wywołania funkcji zastąpić wartością (no bo wywołujemy performPayment), więc mamy efekt uboczny, bla bla bla

Reply:
A może napisać o Future/Promisach?

Chociażby, że można coś zrównoleglić i dopiero po 'tym i tym' powiadomić użytkownika, że to i to.

Choćby taki przykład od Arrow 
```
val speaker = repository.loadSpeaker().await()
val talk = speaker.nextTalk().await()
val conference = talk.getConference().await()
val city = conference.getCity().await()
reservations.bookFlight(speaker, city).await()
```

https://arrow-kt.io/docs/patterns/monads/#deferred-future

Reply:
A to jest już materiał na oddzielny artykuł :) Keep it simple ;)

Reply:
Nie miałem zbytnio pomysłu jak napisać to inaczej, ale coś tam napisałem. Choć znowu wyszło na to, że odbiegłem od tematu. Nie wiem zerknij, czy lepsze. :D